### PR TITLE
Jekyll.sanitized_path: escape tildes before sanitizing a questionable path

### LINKED ISSUE #4457: Files that start with a tilde (~) break builds
The way [we use `File.expand_path`](https://github.com/jekyll/jekyll/blob/7df8f502487f3b6476b5ce0f296be28f70df59cb/lib/jekyll.rb#L156), we expand paths of every file. We should escape a number of different values in this call. [StackOverflow has a great idea of what to escape](http://stackoverflow.com/questions/6159355/how-do-i-escape-parentheses-and-brackets-in-ruby).

```irb
>> File.expand_path("~hi.txt")
ArgumentError: user hi.txt doesn't exist
	from (irb):3:in `expand_path'
	from (irb):3
	from /Users/parkr/.rbenv/versions/2.3.0/bin/irb:11:in `<main>'
>> File.expand_path('\~hi.txt')
=> "/Users/parkr/jekyll/jekyll/\\~hi.txt"
```

Should be patched on 3.0.2 and 3.1.2.

/cc @benbalter

### PR DESCRIPTION
**Pull request by parkr**
Fixes #4457.

/cc @benbalter

### CODE REVIEW

File: lib/jekyll.rb [hunk 1]
```diff
@@ -153,8 +153,9 @@ def sites
     def sanitized_path(base_directory, questionable_path)
       return base_directory if base_directory.eql?(questionable_path)
 
+      questionable_path.insert(0, '/') if questionable_path.start_with?('~')
```
Comment (parkr) [inline lib/jekyll.rb]: @mastahyeti Hello! Could I borrow your :eyes: for a quick moment? We're seeing issues with questionable paths which start with `~`, such as `~$microsoftSwp.docx`. This is an attempt to fix the `ArgumentError`. Is it safe? :lock: 

```irb
>> File.expand_path('~$microsoftSwp.docx', '/')
ArgumentError: user $microsoftSwp.docx doesn't exist
	from (irb):1:in `expand_path'
	from (irb):1
	from /Users/parkr/.rbenv/versions/2.3.0/bin/irb:11:in `<main>'
>> File.expand_path('/~$microsoftSwp.docx', '/')
=> "/~$microsoftSwp.docx"
```

Figured I'd ask you before merging in Jekyll so we can get some extra reassurances. :heart:

**Comment by parkr** [comment]
@benbalter We can't in a PATCH release, but maybe for Jekyll v3.2, [we ignore files that start with `~`](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/entry_filter.rb#L3)?

**Comment by benbalter** [comment]
> @benbalter We can't in a PATCH release, but maybe for Jekyll v3.2, we ignore files that start with ~?

Is there a legitimate use case for publishing files starting for a `~`?

**Comment by parkr** [comment]
> Is there a legitimate use case for publishing files starting for a `~`?

Considering I didn't catch this bug beforehand, I would say "no". If you have `https://example.com/~benbalter/` that usually maps specially for your server to some `/home/benbalter/public_html` folder or something. That's what we did [at McGill CS](http://cs.mcgill.ca/~pmoore6/) way back in the day. So no, generally `~` is considered "special."

File: lib/jekyll.rb (hunk 1 above)
Comment (mastahyeti) [inline lib/jekyll.rb]: Seems fine. Thanks for asking.

**Comment by benbalter** [comment]
:+1:

**Pull request merged**